allen-cell-animated/nucmorph-colorizer
# Feature/url datasets

# Please read! ⬇

Problem
=======
This resolves [#34, pointing to the URL of any hosted dataset](https://github.com/allen-cell-animated/nucmorph-colorizer/issues/34)!

Solution
========
- A new `collection` parameter can be added to the URL, which describes either a path to a directory containing a `collection.json` or a custom `.json` file. This provides names of and relative paths (or URLs) to one or more datasets.
  - The relative paths can either specify a directory containing a `manifest.json`, a path to a custom `.json` file with the same format. They can also be URLs (starting with `http(s)://` or `//`).
  - These dataset names are loaded and then shown in the UI datasets dropdown.
  - If left empty, the collection defaults to the existing nucmorph-colorizer data store.
  
Example `collection.json`:
```
[
    { "name": "Default", "path": "test1" },
    { "name": "Nested", "path": "test2/nested/" },
    { "name": "Custom Manifest", "path": "test3/custom-manifest.json" },
    { "name": "Custom URL", "path": "http://dev-aics-dtp-001.corp.alleninstitute.org/dan-data/colorizer/data/baby_bear"},
]
```
  
- The `dataset` parameter is now overloaded to accept direct URL paths to datasets.
  - If set to a URL, the `collection` parameter is ignored and no other datasets can be selected.
  - The URL can specify either the directory of a `manifest.json` or be a direct path to a custom `.json` file with the same format.
  - If left empty or if the dataset is not in the collection, the first dataset in the collection is loaded by default.
  
- When switching datasets, if the new dataset is missing data (could not be initialized, invalid path, etc.), the last known good dataset will be reloaded instead. (TODO: error popup?)
  
Other changes:
- Added a **timeout fetch** in response to a bug discovered during today's on-premises data outage. The interface would pause while datasets were being fetched, but because the request never resolved, no error message would be printed.
- The current frame number now persists between datasets.


## Type of change
* New feature (non-breaking change which adds functionality)


Steps to Verify:
----------------
- Clone and run the repo with `npm run dev`, which should start the development server at `localhost:5173`.
- All of these URLs should load and show data:

| Collection | Dataset | URL | Note |
| --- | --- | --- | --- |
| None | None | http://localhost:5173 |
| None | Name | http://localhost:5173?dataset=Baby%20Bear |
| None | Base URL | http://localhost:5173?dataset=http%3A%2F%2Fdev-aics-dtp-001.corp.alleninstitute.org%2Fdan-data%2Fcolorizer%2Fdata%2Fbaby_bear |
| None | JSON URL | http://localhost:5173/?dataset=http%3A%2F%2Fdev-aics-dtp-001.corp.alleninstitute.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Ftest3%2Fcustom-manifest.json |
| Base URL |  None | http://localhost:5173/?collection=http%3A%2F%2Fdev-aics-dtp-001.corp.alleninstitute.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest |
| JSON URL | None | http://localhost:5173/?collection=http%3A%2F%2Fdev-aics-dtp-001.corp.alleninstitute.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Ftest.json  | Should say CUSTOM in dataset box. Also, try loading dataset 4 to see default behavior for bad datasets. |
| Base URL |  Name | http://localhost:5173/?collection=http%3A%2F%2Fdev-aics-dtp-001.corp.alleninstitute.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest&dataset=Test%202%20Nested | Should load Test 2 on startup |


Screenshots (optional):
-----------------------
![image](https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/b0d6a1af-80a3-4d10-9f18-f1f969c140f5)


Thanks for contributing!
